cli/src/diff_util.rs

```diff
@@ -644,6 +681,67 @@ pub fn show_color_words_diff(
     Ok(())
 }
 
+pub fn show_file_by_file_diff(
+    ui: &Ui,
+    repo: &dyn Repo,
+    formatter: &mut dyn Formatter,
+    tool: &ExternalMergeTool,
+    tree_diff: TreeDiffStream,
+    path_converter: &RepoPathUiConverter,
+) -> Result<(), DiffRenderError> {
+    fn create_file(
+        path: &RepoPath,
+        wc_dir: &Path,
+        value: MaterializedTreeValue,
+    ) -> Result<PathBuf, DiffRenderError> {
+        let fs_path = path.to_fs_path(wc_dir);
+        std::fs::create_dir_all(fs_path.parent().unwrap())?;
+        let content = diff_content(path, value)?;
+        let mut buffer = File::create(&fs_path)?;
+        buffer.write_all(&content.contents)?;
+        Ok(fs_path)
+    }
+
+    let temp_dir = new_utf8_temp_dir("jj-diff-")?;
+    let left_wc_dir = temp_dir.path().join("left");
+    let right_wc_dir = temp_dir.path().join("right");
+    let mut diff_stream = materialized_diff_stream(repo.store(), tree_diff);
+    async {
+        while let Some((path, diff)) = diff_stream.next().await {
+            let ui_path = path_converter.format_file_path(&path);
+            let (left_value, right_value) = diff?;
+
+            match (&left_value, &right_value) {
+                (_, MaterializedTreeValue::AccessDenied(source))
+                | (MaterializedTreeValue::AccessDenied(source), _) => {
+                    write!(
+                        formatter.labeled("access-denied"),
+                        "Access denied to {ui_path}:"
+                    )?;
+                    writeln!(formatter, " {source}")?;
+                    continue;
+                }
+                _ => {}
+            }
+            let left_path = create_file(&path, &left_wc_dir, left_value)?;
+            let right_path = create_file(&path, &right_wc_dir, right_value)?;
+
+            invoke_external_diff(
+                ui,
+                formatter.raw(),
+                tool,
+                maplit::hashmap! {
+                    "left" => left_path.to_str().expect("temp_dir should be valid utf-8"),
+                    "right" => right_path.to_str().expect("temp_dir should be valid utf-8"),
+                },
+            )
+            .map_err(DiffRenderError::DiffGenerate)?;
```

Comment:
Can't we reuse `check_out_trees()` to materialize files all at once, and execute diff tool for each file pair? That will help extend file-by-file mode to something like `tool -- file1a file1b -- file2a file2b` or `tool file1a file1b &; tool file2a file2b &;`.
https://github.com/martinvonz/jj/issues/3696#issuecomment-2116442540

Reply:
I want to avoid a full checkout because the way I plan to do file copy tracking so that it works for all our tools is in the materialize method we are using.

Reply:
Can't we port `check_out_trees()` to be based on the diff materialization logic then? It'll be nice if `diffedit --tool` can somehow deal with file renames.

Reply:
I don't think so because the trees need to have files with actual names in them that makes such a mapping difficult.  How would you represent a copy cycle of A -> B -> C ->A for example.

Reply:
Maybe I don't understand the underlying problem, but `check_out_trees()` first calculates `left_tree.diff_stream(right_tree, matcher)` to build a sparse pattern, then checks out files from `left`/`right_trees`. I assumed it's roughly equivalent to dump file pairs yielded by `materialized_diff_stream()`.

Reply:
when you create the directory structure it most represent both sides for all file simulatenously.  The stream only has to represent both sides for each file one at a time.  As a result, a stream can handle oddities like cycles of renames where the directory cannot.

Reply:
I don't quite understand the reasoning, sorry. Perhaps, this part should be reviewed by someone else who have expertise on copy tracing.

> The stream only has to represent both sides for each file one at a time. As a result, a stream can handle oddities like cycles of renames where the directory cannot.

Does it mean we can't support `files-all-at-once` mode (in which jj will materialize all changed file pairs and spawn diff tool with file name pairs e.g. `tool a/foo b/foo &; tool a/bar b/renamed-bar &; ...`)?

Reply:
I also am a little confused about why we couldn't just make two directories with all the correct files (but I'm also not an expert on copy tracking). If a rename cycle is so complicated that the "before" and "after" states cannot be represented correctly, perhaps we should prohibit it entirely?

Do you have an example where it's a problem?

As an aside, I don't think this PR is a prerequisite for copy tracking, or at least I'm missing the reason why it would be. At first, it would be OK if only the internal diff knew how to pair the files.

Reply:
At a high level, the directory approach doesn't have a good mechanism for copy tracking because there is no way to communicate to the external tool what the mapping of files is.  So you have two options:

1. lay out the directories as they actually exist at each version and then be unable to tell the tool what the copies are (status quo)
2. lay out the directories using a set of names that moves around old files based on where the new files actually end up so that external tool will show the right thing (won't properly handle copy cycles and also feels error prone on the logic side) 

If you have a thing where to files swap names (easier to think about then the three way cycle).  how do you suggest laying out the directory structure?  I am fairly certain there is not a simple answer that will make it work.

The PR is not a hard prereq to using copy tracking in any diff stuff, but it is a pre-req to external diff tools support for copy tracking given the above.  Since my explicit interest in this is with an external diff tool, I don't feel bad about pushing it because of that.

Reply:
> If you have a thing where to files swap names (easier to think about then the three way cycle). how do you suggest laying out the directory structure?

Let's X was renamed to Y and Y was renamed to X.Let's say that happened in commit B, which is on top of commit A. Then I would expect a diff files X and Y, where the LHS of X is Y@A and the RHS isX@B, and where the LHS of Y is X@A and the the RHS is Y@B. Should be possible, right?

It can still be confusing to the user. To help with that, we can print a message to the terminal, or maybe to a temporary file (like the `JJ-INSTRUCTIONS` file we already have).

Reply:
the only way to accomplish what you are suggesting in the directory structure is to have the thing in LEFT not be an actual checkout but to have it be a checkout with some renames applied, which is weird IMO.

File-by-file on the other hand this is a very natural representation.

Reply:
I was thinking of `file-by-file` making the same checkout as normal, and then calling the tool file-by-file. Future methods could provide the pairings to the tool in other ways (pairs of arguments, text file).

Even for dirdiff, you can describe the correspondence in JJ-INSTRUCTIONS, as I think Martin suggested just above.


> lay out the directories as they actually exist at each version and then be unable to tell the tool what the copies are (status quo)

For the minimal dirs, there is some advantage move/copy tracking can provide you. If file A was copied to B, you can put A on the LHS and B on the RHS (as opposed to putting nothing on the LHS as would happen now).

Reply:
that will end up replicating a lot of the logic for copy detection in a way that feel awkward to me.  I think the proper place to isolate that logic is inside materialize_diff_stream which will allow for common handling across all our diff types

Reply:
Are you trying to get this PR into the Wednesday release?

In that case, if we can't get to the same page, one possibility would be to add `experimental-` to the option name, expecting future breaking changes (if Martin/Yuya agree).

As far as I'm concerned, I don't want to block your progress, but I'm not sure there is a consensus here. I'm, personally, not entirely opposed to creating files one-by-one, but I still don't understand the decision to do that very well (it could be just a matter of time).

If Martin or Yuya are on the same page as you, I'm OK with this.

Reply:
no, I am happy to miss the release since the thing I really want is the integration that is a next step

Reply:
I have a pretty clear idea in my mind of how I want to use this, so I think it will become more clear in a follow up CL when I integrate these things into the copy detection work

Reply:
> I was thinking of `file-by-file` making the same checkout as normal, and then calling the tool file-by-file. Future methods could provide the pairings to the tool in other ways (pairs of arguments, text file).

I was considering this, too. If the diff tool is GUI/CUI-based, user might want to open all pairs at once.

And I assume the implementation can be switched to be based off `materialize_diff_stream()` if that makes sense, something like:
```rust
let mut file_name_pairs: Vec<(RepoPath, RepoPath)>;
while let Some(((left_path, right_path), (left_value, right_value))) = diff_stream.next().await {
    write_file(left_temp_dir, left_path, left_value);
    write_file(right_temp_dir, right_path, right_value);
    file_name_pairs.push((left_path, right_path));
}
if diffedit {
    somehow_construct_tree_state_for_snapshot(right_temp_dir, file_name_pairs); // XXX
}
for (left_path, right_path) in file_name_pairs {
    invoke_external_tool(..);
    if !batch {
        wait();
    }
}
```

Reply:
deferring this discussion until later work that integrates copy-tracking.  I think that will make it more clear what the best answer is